File: src/com/dylowen/tinglebot/brain/Brain.java
package com.dylowen.tinglebot.brain;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import com.dylowen.tinglebot.Timer;

/**
 * T the word type
 * V the sentence type
 *
 * @author dylan.owen
 * @since Feb-2016
 */
public abstract class Brain<T, V>
    implements Serializable {

    protected transient static final int MIN_GRAM_SIZE = 2;

    protected final Map<NGram<T>, WeightedSet<T>> dictionary = new HashMap<>();
    protected int gramSize;

    private transient Random rand = new Random();
    private transient Map.Entry<NGram<T>, WeightedSet<T>>[] cachedEntryArray = null;
    private transient LinkedList<T> input = null;

    public Brain(final int gramSize) {
        this.gramSize = gramSize;
    }

    public void feed(final T word) {
        if (this.input == null) {
            this.input = new LinkedList<>();
        }
        this.input.add(word);
        //System.out.println(word);

        //not enough input
        if (this.input.size() <= this.gramSize) {
            return;
        }

        //create the NGrams
        for (int i = 0; i <= this.gramSize - MIN_GRAM_SIZE; i++) {
            final List<T> subList = this.input.subList(i, this.input.size() - 1);
            final NGram<T> nGram = new NGram<>(subList);

            WeightedSet<T> set = this.dictionary.get(nGram);
            if (set == null) {
                set = new WeightedSet<>();

                this.dictionary.put(nGram, set);
            }

            set.add(word);

            //System.out.println(nGram.toString() + " [" + set.toString() + "]");
        }

        this.input.removeFirst();
    }

    public void compress() {}

    public V getSentence() {
        final List<T> words = getSentenceWords();

        return concatSentence(words);
    }

    public abstract V concatSentence(final List<T> words);

    public List<T> getSentenceWords() {
        final Map.Entry<NGram<T>, WeightedSet<T>> first = getRandomEntry();

        final LinkedList<T> sentence = new LinkedList<T>(first.getKey().getWords());
        continueSentence(sentence);

        return sentence;
    }

    public List<T> getSentenceWords(final LinkedList<T> input) {
        final LinkedList<T> sentence = new LinkedList<>(input.subList(input.size() - this.gramSize, input.size()));

        continueSentence(sentence);

        //strip out our input
        return sentence.subList(this.gramSize - 1, sentence.size());
    }

    private void continueSentence(final LinkedList<T> sentence) {
        T nextWord;
        do {
            nextWord = getNextWord(sentence);
            sentence.add(nextWord);
        } while (shouldContinueSentence(nextWord, sentence));
        //System.out.println(GenericWordType.getType(nextWord).name() + " " + nextWord);
    }

    protected abstract boolean shouldContinueSentence(final T nextWord, final LinkedList<T> sentence);

    protected T getNextWord(final LinkedList<T> input) {
        final int startIndex = (this.gramSize > input.size()) ? 0 : input.size() - this.gramSize;
        final List<T> operatingList = new LinkedList<>(input.subList(startIndex, input.size()));

        WeightedSet<T> set;
        do {
            final NGram operatingGram = new NGram<>(operatingList);
            set = this.dictionary.get(operatingGram);

            //System.out.println(operatingGram.toString() + ((set != null) ? " [" + set.toString() + "]" : ""));
            //System.out.println(operatingList.size() + " " + ((set != null) ? set.size() + " " : "")
            //        + (operatingList.size() > MIN_GRAM_SIZE) + " " + (set == null || set.size() <= 1));

            operatingList.remove(0);
            //loop if we get a null set or the set size is uninteresting and we still have an operating list
        } while (operatingList.size() > MIN_GRAM_SIZE - 1 && (set == null || set.size() <= 1));

        //we've run out of stuff to say
        if (set == null) {
            return null;
        }

        return set.get();
    }

    //this could be better
    private Map.Entry<NGram<T>, WeightedSet<T>> getRandomEntry() {
        if (this.cachedEntryArray == null || this.cachedEntryArray.length != this.dictionary.size()) {

            final Set<Map.Entry<NGram<T>, WeightedSet<T>>> entrySet = this.dictionary.entrySet();
            //TODO can I get rid of this unchecked?
            @SuppressWarnings("unchecked")
            final Map.Entry<NGram<T>, WeightedSet<T>>[] entryArray = new Map.Entry[entrySet.size()];

            this.cachedEntryArray = entrySet.toArray(entryArray);
        }

        Map.Entry<NGram<T>, WeightedSet<T>> entry;
        do {
            entry = this.cachedEntryArray[this.rand.nextInt(this.cachedEntryArray.length)];
        } while (entry.getKey().getWords().size() < this.gramSize);

        return entry;
    }

    public int stateCount() {
        return this.dictionary.size();
    }

    public void export(final String exportPath) {
        final Timer timer = new Timer();

        try {
            FileOutputStream fileOut = new FileOutputStream(exportPath);
            ObjectOutputStream out = new ObjectOutputStream(fileOut);
            out.writeObject(this);
            out.close();
            fileOut.close();
            System.out.println("Serialized brain is saved in: " + exportPath);
        }
        catch (IOException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

        System.out.println("Brain export time: " + timer.getS() + "s");
    }

    private void readObject(final ObjectInputStream stream)
            throws IOException, ClassNotFoundException {
        stream.defaultReadObject();
        this.rand = new Random();
    }
}

File: src/main/scala/com/dylowen/tinglebot/brain/Brain.scala
package com.dylowen.tinglebot.brain

import scala.annotation.tailrec
import scala.util.Random

/**
  * TODO add description
  *
  * @author dylan.owen
  * @since Mar-2016
  */
object Brain {
  val MIN_GRAM_SIZE: Integer = 2
}

abstract class Brain[T, V](protected var gramSize: Integer) extends Serializable {
  protected var dictionary: Map[NGram[T], WeightedSet[T]] = Map()

  //val dictionary: scala.collection.mutable.Map[NGram[T], WeightedSetJava[T]] = new mutable.HashMap[]()

  @transient
  private val rand: Random = new Random()
  @transient
  private var input: List[T] = Nil

  def feed(word: T): Unit = {
    if (this.input.size >= this.gramSize) {
      var subList: List[T] = this.input
      for (i <- 0 to this.gramSize - Brain.MIN_GRAM_SIZE) {
        val nGram = new NGram[T](subList)

        val set = this.dictionary.get(nGram) match {
          case Some(v) => v
          case None =>
            val newSet = new WeightedSet[T]()
            this.dictionary += (nGram -> newSet)

            newSet
        }

        set.add(word)

        subList = subList.tail
      }

      this.input = this.input.tail
    }

    this.input :+= word
  }

  def compress(): Unit

  def getSentenceWords: List[T] = {
    val start: List[T] = this.getRandomWords

    continueSentence(start)
  }

  @tailrec
  private def continueSentence(sentenceHead: List[T]): List[T] = {
    val nextWord: Option[T] = getNextWord(sentenceHead)

    if (nextWord.exists(v => !shouldContinueSentence(v, sentenceHead))) {
      return sentenceHead
    }

    continueSentence(sentenceHead :+ nextWord.get)
  }

  protected def shouldContinueSentence(word: T, sentence: List[T]): Boolean

  protected def getNextWord(input: List[T]): Option[T] = {
    val startIndex = if (this.gramSize > input.size) 0 else input.size - this.gramSize
    var operatingList = input.slice(startIndex, input.size)

    var set: Option[WeightedSet[T]] = None
    do {
      val operatingGram: NGram[T] = new NGram[T](operatingList)
      set = this.dictionary.get(operatingGram)

      //System.out.println(operatingGram.toString() + " : [" + set.map(_.toString).getOrElse("") + "]")
      //System.out.println(operatingList.size + " : [" + set.map(_.size).getOrElse("") + "]\n")

      operatingList = operatingList.tail
      //loop if we get a null set or the set size is uninteresting and we still have an operating list
    } while (operatingList.size > Brain.MIN_GRAM_SIZE - 1 && (set.isEmpty || set.exists(v => v.size <= 1)))

    set.map(v => v.get)
  }

  def concatSentence(word: List[T]): V

  def stateCount = this.dictionary.size

  private var cachedNGramArray: Array[NGram[T]] = null
  private def getRandomWords: List[T] = {
    if (this.cachedNGramArray == null || this.cachedNGramArray.length != this.dictionary.size) {
      this.cachedNGramArray = this.dictionary.keySet.toArray
    }

    while (true) {
      val startGram: NGram[T] = this.cachedNGramArray(this.rand.nextInt(this.cachedNGramArray.length))

      if (startGram.size >= this.gramSize) {
        return startGram.words
      }
    }

    throw new AssertionError("You should never get here")
  }
}
